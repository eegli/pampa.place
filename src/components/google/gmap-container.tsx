import {MapService} from '@/services/google';
import Box from '@mui/material/Box';
import {ReactNode, useEffect, useRef} from 'react';

type InstanceMethods = Omit<
  InstanceType<typeof google.maps.Map>,
  'setCenter' | 'fitBounds'
>;

export type GoogleMapContainerProps = {
  children: ReactNode;
  onMount?: (map: InstanceMethods) => unknown;
  onUnmount?: (map: InstanceMethods) => unknown;
} & (
  | {
      bounds: google.maps.LatLngLiteral;
      padding?: number;
    }
  | {
      center: google.maps.LatLngLiteral;
      zoom?: number;
    }
);

export const GoogleMapContainer = ({
  children,
  onMount,
  onUnmount,
  ...rest
}: GoogleMapContainerProps) => {
  const ref = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (ref.current) {
      const unmount = MapService.mount(ref.current);

      if ('bounds' in rest) {
        /* Order in constructor is important! SW, NE  */
        const bounds = new google.maps.LatLngBounds(rest.bounds);
        google.maps.event.addListenerOnce(MapService.map, 'idle', () => {
          MapService.map.fitBounds(bounds, rest.padding || 0);
        });
      } else {
        MapService.map.setCenter(rest.center);
        MapService.map.setZoom(rest.zoom || 0);
      }

      if (typeof onMount === 'function') {
        onMount(MapService.map);
      }

      return () => {
        if (typeof onUnmount === 'function') {
          onUnmount(MapService.map);
        }
        unmount();
      };
    }
  }, [ref, onMount, onUnmount, rest]);

  return (
    <>
      <Box data-testid="play-google-map" ref={ref} height="100%" width="100%" />
      {children}
    </>
  );
};
