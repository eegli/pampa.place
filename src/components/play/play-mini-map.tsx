import CloseIcon from '@mui/icons-material/Close';
import OpenInNewIcon from '@mui/icons-material/OpenInNew';
import {Box, Button, IconButton} from '@mui/material';
import {forwardRef, ReactNode} from 'react';

type MiniMapProps = {
  enlarged: boolean;
  isPositionSelected: boolean;
  successCallback: () => void;
  toggle: () => void;
  children: ReactNode;
};

export const MiniMap = forwardRef<HTMLDivElement, MiniMapProps>(
  function MiniMapWithRef(props, ref) {
    const {enlarged, isPositionSelected, successCallback, toggle, children} =
      props;

    return (
      <Box
        ref={ref}
        id="mini-map"
        role="region"
        position="absolute"
        bottom={50}
        right={30}
        maxHeight="70%"
        maxWidth="85%"
        minHeight="20%"
        minWidth="30%"
        height={enlarged ? 700 : 150}
        width={enlarged ? 700 : 200}
        sx={{
          transition: '0.2s ease',
        }}
        // Display on top of speed dial
        zIndex={1000}
      >
        <IconButton
          aria-label="mini-map open button"
          onClick={toggle}
          size="large"
          sx={{
            position: 'absolute',
            zIndex: 1000,
            display: enlarged ? 'none' : 'block',
            height: '100%',
            width: '100%',
            background:
              'linear-gradient(rgba(0, 0, 0, 0.6), rgba(0, 0, 0, 0.6))',
            borderRadius: 0,
          }}
        >
          <OpenInNewIcon fontSize="large" />
        </IconButton>
        {children}

        {enlarged ? (
          <>
            <IconButton
              aria-label="mini-map close button"
              onClick={toggle}
              size="large"
              sx={{
                position: 'absolute',
                top: '2%',
                right: '2%',
                background:
                  'linear-gradient(rgba(0, 0, 0, 0.4), rgba(0, 0, 0, 0.4))',
              }}
            >
              <CloseIcon fontSize="large" />
            </IconButton>
            <Button
              aria-label="location submit button"
              variant="contained"
              color={isPositionSelected ? 'success' : 'secondary'}
              onClick={successCallback}
              sx={{
                position: 'absolute',
                bottom: 0,
                left: 0,
                width: '100%',
              }}
            >
              {isPositionSelected ? "I'm here!" : 'Place the marker'}
            </Button>
          </>
        ) : null}
      </Box>
    );
  }
);
