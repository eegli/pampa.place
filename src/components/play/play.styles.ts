import { Box, BoxProps, styled } from '@mui/material';

interface StyledMapOverlayProps extends BoxProps {
  pos: 'map' | 'submit';
}
export const StyledMapOverlay = styled(Box)<StyledMapOverlayProps>(
  ({ pos }) => ({
    zIndex: 10,
    position: 'fixed',
    width: '4rem',
    bottom: pos === 'map' ? 40 : 130,
    right: pos === 'map' ? 70 : 90,
    '&:hover': {
      cursor: 'pointer',
    },
  })
);
