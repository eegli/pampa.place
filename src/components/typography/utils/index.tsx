import Box from '@mui/material/Box';

export const em = (txt: string, color: 'g' | 'v') => (
  <Box
    component="span"
    color={color === 'g' ? 'success.main' : 'secondary.main'}
  >
    <b>{txt}</b>
  </Box>
);
