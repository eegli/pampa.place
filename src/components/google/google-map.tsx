import {MapService} from '@/services/google';
import Box from '@mui/material/Box';
import {memo, ReactNode, useEffect, useRef} from 'react';

type BoundsProps = {
  bounds: {
    SW: google.maps.LatLngLiteral;
    NE: google.maps.LatLngLiteral;
  };
  padding?: number;
};

type CenterProps = {
  center: google.maps.LatLngLiteral;
  zoom?: number;
};

export type GoogleMapProps = {
  id?: string;
  children?: ReactNode;
  onMount?: (map: google.maps.Map) => unknown;
  onUnmount?: (map: google.maps.Map) => unknown;
} & (BoundsProps | CenterProps);

const _GoogleMap = ({
  children,
  id = 'google-map',
  onMount,
  onUnmount,
  ...rest
}: GoogleMapProps) => {
  const ref = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (ref.current) {
      const unmountMap = MapService.mount(ref.current);
      console.info('%cGOOGLE MAP MOUNT', 'color: yellow');

      if ('bounds' in rest) {
        /* Order in constructor is important! SW, NE  */
        const bounds = new google.maps.LatLngBounds(
          rest.bounds.SW,
          rest.bounds.NE
        );
        google.maps.event.addListenerOnce(MapService.map, 'idle', () => {
          MapService.map.fitBounds(bounds, rest.padding || 0);
        });
      } else {
        MapService.map.setCenter(rest.center);
        MapService.map.setZoom(rest.zoom || 0);
      }

      if (typeof onMount === 'function') {
        onMount(MapService.map);
      }

      return () => {
        // Any cleanup logic
        if (typeof onUnmount === 'function') {
          onUnmount(MapService.map);
        }

        unmountMap();
        console.info('%cGOOGLE MAP UNMOUNT', 'color: yellow');
      };
    }
  }, [ref, onMount, onUnmount, rest]);

  return (
    <Box id={id} data-testid={id} ref={ref} height="100%" width="100%">
      {children}
    </Box>
  );
};

export const GoogleMap = memo(_GoogleMap, (prev, next) => prev.id === next.id);
