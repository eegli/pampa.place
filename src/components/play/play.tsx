import {GoogleMap} from '@/components/google/map';
import {GoogleStreetView} from '@/components/google/street-view';
import {MAPS} from '@/config/maps';
import {setPlayerScore} from '@/redux/game';
import {getActivePlayer} from '@/redux/game/selectors';
import {useAppDispatch, useAppSelector} from '@/redux/hooks';
import {Box, Button} from '@mui/material';
import {BoxProps} from '@mui/material/Box';
import {styled} from '@mui/material/styles';
import {useState} from 'react';
import {PlayHeader} from './play-header';

interface StyledMapOverlayProps extends BoxProps {
  pos: 'map' | 'submit';
}
const StyledMapOverlay = styled(Box)<StyledMapOverlayProps>(({pos}) => ({
  zIndex: 10,
  position: 'fixed',
  width: '4rem',
  bottom: pos === 'map' ? 40 : 130,
  right: pos === 'map' ? 70 : 90,
  '&:hover': {
    cursor: 'pointer',
  },
}));

export const Play = () => {
  const dispatch = useAppDispatch();
  const [showMap, setShowMap] = useState<boolean>(false);

  const activePlayer = useAppSelector(getActivePlayer);
  const initialPos = useAppSelector(({position}) => position.initialPosition);
  const selectedPos = useAppSelector(({position}) => position.selectedPosition);
  const activeMapId = useAppSelector(({game}) => game.mapId);
  const map = MAPS.get(activeMapId);

  function submitScore() {
    dispatch(
      setPlayerScore({
        initial: initialPos,
        selected: selectedPos,
      })
    );
  }

  if (!map || !initialPos) {
    return null;
  }

  return (
    <Box
      id="play-mode"
      sx={{
        display: 'flex',
        flexFlow: 'column',
        alignItems: 'center',
        height: '100%',
        width: '100%',
        overflow: 'hidden',
      }}
    >
      <PlayHeader player={activePlayer} timerCallback={submitScore} />
      <Box position="relative" height="100%" width="100%">
        <StyledMapOverlay pos="map" onClick={() => setShowMap(!showMap)}>
          {/* eslint-disable-next-line  @next/next/no-img-element */}
          <img aria-label="map-toggle" src="/map.svg" alt="map-icon" />
        </StyledMapOverlay>

        <StyledMapOverlay pos="submit" onClick={() => setShowMap(!showMap)}>
          <Button
            size={'large'}
            variant="contained"
            color="secondary"
            onClick={submitScore}
            disabled={!selectedPos}
          >
            Submit
          </Button>
        </StyledMapOverlay>
        <GoogleStreetView display={showMap ? 'none' : 'block'} />
        <GoogleMap map={map} mode="play" />
      </Box>
    </Box>
  );
};
