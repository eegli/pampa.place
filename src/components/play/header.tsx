import {useAppSelector} from '@/redux/hooks';
import {formatDur} from '@/utils/misc';
import {Box, Divider, Stack, Typography} from '@mui/material';
import {useRouter} from 'next/router';
import {useEffect} from 'react';
import {useTimer} from '../../hooks/useTimer';

type PlayHeaderProps = {
  player: string;
  timerCallback: () => void;
};

export const PlayHeader = ({player, timerCallback}: PlayHeaderProps) => {
  const router = useRouter();
  const timeLimit = useAppSelector(s => s.game.timeLimit);
  const [timeRemaining] = useTimer(timeLimit);

  const isUnlimitedTimeMode = timeLimit < 0;

  const label = isUnlimitedTimeMode ? (
    <span>&#8734;</span>
  ) : (
    <span>{formatDur(timeRemaining)}</span>
  );

  useEffect(() => {
    if (!isUnlimitedTimeMode && !timeRemaining) {
      timerCallback();
    }
  }, [timeRemaining, isUnlimitedTimeMode, timerCallback]);

  function handleClick() {
    router.push('/');
  }

  return (
    <Box px={2} my={2} width="100%" display="flex">
      <Stack
        flexGrow={1}
        display="flex"
        justifyContent="center"
        direction="row"
        divider={<Divider orientation="vertical" flexItem />}
        spacing={2}
      >
        <Box textAlign="right">
          <Typography
            variant="caption"
            color={'text.secondary'}
            sx={{
              textTransform: 'uppercase',
            }}
          >
            Your turn
          </Typography>
          <Typography variant="h5">{player}</Typography>
        </Box>
        <Box>
          <Typography
            variant="caption"
            color={'text.secondary'}
            sx={{
              textTransform: 'uppercase',
            }}
          >
            Time remaining
          </Typography>
          <Typography variant="h5" color={'secondary.main'}>
            {label}
          </Typography>
        </Box>
      </Stack>
    </Box>
  );
};
