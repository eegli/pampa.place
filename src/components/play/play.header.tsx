import { useAppSelector } from '@/redux/hooks';
import { formatDuration } from '@/utils/misc';
import HomeIcon from '@mui/icons-material/Home';
import { Box, Divider, IconButton, Stack, Typography } from '@mui/material';
import { useRouter } from 'next/router';
import { useEffect } from 'react';
import { useTimer } from '../../hooks/useTimer';

type PlayHeaderProps = {
  player: string;
  timerCallback: () => void;
};

const getLabel = (unlimited: boolean, time: number) => {
  return unlimited ? <span>&#8734;</span> : <span>{formatDuration(time)}</span>;
};

const PlayHeader = ({ player, timerCallback }: PlayHeaderProps) => {
  const router = useRouter();
  const timeLimit = useAppSelector(s => s.game.timeLimit);
  const [timeRemaining] = useTimer(timeLimit);

  const isUnlimitedTimeMode = timeLimit < 0;

  const label = getLabel(isUnlimitedTimeMode, timeRemaining);

  useEffect(() => {
    if (!timeRemaining && !isUnlimitedTimeMode) {
      timerCallback();
    }
  }, [timeRemaining, isUnlimitedTimeMode, timerCallback]);

  function handleClick() {
    router.push('/');
  }

  return (
    <Box px={2} my={2} width="100%" display="flex">
      <Stack
        flexGrow={1}
        display="flex"
        justifyContent="center"
        direction="row"
        divider={<Divider orientation="vertical" flexItem />}
        spacing={2}
      >
        <Box textAlign="right">
          <Typography
            variant="caption"
            color={'text.secondary'}
            sx={{
              textTransform: 'uppercase',
            }}
          >
            Your turn
          </Typography>
          <Typography variant="h5">{player}</Typography>
        </Box>
        <Box>
          <Typography
            variant="caption"
            color={'text.secondary'}
            sx={{
              textTransform: 'uppercase',
            }}
          >
            Time remaining
          </Typography>
          <Typography variant="h5" color={'secondary.main'}>
            {label}
          </Typography>
        </Box>
      </Stack>

      <IconButton sx={{ mx: 2 }} size="large" edge="end" onClick={handleClick}>
        <HomeIcon />
      </IconButton>
    </Box>
  );
};
export default PlayHeader;
