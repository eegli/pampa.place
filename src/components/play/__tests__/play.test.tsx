import {updateSelectedPosition} from '@/redux/position';
import {
  createMockState,
  createMockStore,
  render,
  screen,
} from '@/tests/test-utils';
import React from 'react';
import {mocked} from 'ts-jest/utils';
import {MapService} from '../../../services/google';
import {Play} from '../play';

/* WORK IN PROGRESS */

const mockGoogle = mocked(google.maps, true);
const mockGmap = mocked(MapService, true);

const removeEventListener = jest.fn();

jest.spyOn(mockGmap.map, 'addListener').mockImplementation((event, handler) => {
  return {remove: removeEventListener};
});

jest
  .spyOn(mockGoogle.event, 'addListenerOnce')
  .mockImplementation((_, event, handler) => {
    return {remove: removeEventListener};
  });

afterEach(() => {
  jest.clearAllMocks();
});

describe('Play page', () => {
  it('renders', () => {
    const state = createMockState();
    const store = createMockStore(state);
    const {container} = render(<Play />);
    let button = screen.queryByRole('button', {name: /submit/i});
    expect(button).toBeDisabled();
    store.dispatch(updateSelectedPosition);
    button = screen.queryByRole('button', {name: /submit/i});

    expect(container).toMatchSnapshot();

    expect(button).toBeInTheDocument();
  });
});
