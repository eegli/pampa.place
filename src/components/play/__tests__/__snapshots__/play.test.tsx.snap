// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Play page renders 1`] = `
<div>
  <div
    class="MuiContainer-root css-vg1u2y-MuiContainer-root"
    id="mother"
  >
    <div
      data-testid="__GMAP__CONTAINER__"
      id="__GMAP__CONTAINER__"
    />
    <div
      data-testid="__GSTV__CONTAINER__"
      id="__GSTV__CONTAINER__"
    >
      <div
        data-testid="__GSTV__"
        id="__GSTV__"
        style="height: 100%;"
      />
    </div>
    <div
      class="MuiBox-root css-kgpjue"
    >
      <div
        class="MuiBox-root css-10dfrhn"
      >
        <div
          class="css-197pqds-MuiStack-root"
        >
          <div
            class="MuiBox-root css-s2uf1z"
          >
            <span
              class="MuiTypography-root MuiTypography-caption css-1vsy5xt-MuiTypography-root"
            >
              Your turn
            </span>
            <h5
              class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
            />
          </div>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-1kpnj7b-MuiDivider-root"
          />
          <div
            class="MuiBox-root css-0"
          >
            <span
              class="MuiTypography-root MuiTypography-caption css-1vsy5xt-MuiTypography-root"
            >
              Time remaining
            </span>
            <h5
              class="MuiTypography-root MuiTypography-h5 css-1v2udv1-MuiTypography-root"
            >
              <span>
                1m
              </span>
            </h5>
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root css-ojqyia"
      >
        <div
          class="MuiBox-root css-aj37i8"
        >
          <img
            alt="map-icon"
            src="/map.svg"
          />
        </div>
        <div
          class="MuiBox-root css-1dmpwk8"
        >
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonBase-root Mui-disabled css-1npud8i-MuiButtonBase-root-MuiButton-root"
            disabled=""
            tabindex="-1"
            type="button"
          >
            Submit
          </button>
        </div>
        <div
          style="display: block; height: 100%;"
        />
        <div
          style="height: 100%;"
        >
          <div
            data-testid="__GMAP__"
            id="__GMAP__"
            style="height: 100%;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
